File: 코테_연습/프로그래머스/힙/이중우선순위큐.py
import heapq

def solution(operations):
    heap = []
    heapq.heapify(heap)
    answer = []

    for operation in operations:
        i, number = operation.split(' ')
        number = int(number)

        if i == 'I':
            heapq.heappush(heap, number)
        elif i == 'D':
            if len(heap) == 0: continue
            if number == 1:
                heap = list(heap)[:-1]
                heapq.heapify(heap)
            elif number == -1:
                heapq.heappop(heap)

    if len(heap) == 0:
        return [0, 0]
    else:
        smallest = heap[0]
        largest = sorted(heap)[-1]
        answer = [largest, smallest]
    return answer

operations = ["I -45", "I 653", "D 1", "I -642", "I 45", "I 97", "D 1", "D -1", "I 333"]
print(solution(operations))

File: 코테_연습/힙/프로그래머스/이중우선순위큐.js
function solution(operations) {
  const answer = []

  operations.forEach(operation => {
    let [order, number] = operation.split(' ')
    number = parseInt(number)

    if (order === 'I') {
      answer.push(number)
    } else {
      if (!answer.length) return
      let val = (number === -1 ? Math.min : Math.max)(...answer)
      let idx = answer.findIndex(value => value === val)

      answer.splice(idx, 1)
    }
  })

  return answer.length ? [Math.max(...answer), Math.min(...answer)] : [0, 0]
}

operations = ["I -45", "I 653", "D 1", "I -642", "I 45", "I 97", "D 1", "D -1", "I 333"]
console.log(solution(operations))